c7n/policy.py

```diff
@@ -207,14 +207,14 @@ def provision(self):
 execution = PluginRegistry('c7n.execution')
 
 
-@execution.register('pull')
-class PullMode(PolicyExecutionMode):
-    """Pull mode execution of a policy.
+class ResourceBasedMode(PolicyExecutionMode):
+    """Abstract execution mode of a policy.
 
-    Queries resources from cloud provider for filtering and actions.
+    Children must override list_resources method so that it returns a list of dictionaries.
```

Comment:
the changes to this module don't seem to have any clear reason. child resources use a separate query resource manager and source, not a specialized execution mode.

Reply:
We are going to implement 'create' resource actions later and currently we do not see an appropriate mode to support them. PullMode that is close enough and chosen by default won't execute an action unless there is an existing resource. As generally PullMode's logic works, I have extracted it to an abstract ResourceBasedMode class so that PullMode becomes its child and provides a list of resources taken from API. Besides, in order to support 'create' action that was mentioned at the beginning, I have added PushMode that returns a single empty dictionary which is exactly what is needed in order to create a new single resource.
If there is an existing approach of creating resources, it would be great if you could advise.

Reply:
creating what resources? custodian isn't a provisioning tool for applications. please start with an issue describe the use case.

Reply:
The initial use case was backing up a sql database with https://cloud.google.com/sql/docs/mysql/admin-api/v1beta4/backupRuns/insert, possibly there could be more. As that seems to conflict with the purpose of Custodian, we will be thinking of another way.